tensorflow/contrib/go/graph.go

```diff
@@ -0,0 +1,570 @@
+func (gr *Graph) String() string {
+	var bufStr bytes.Buffer
+	proto.MarshalText(&bufStr, gr.def)
+
+	return bufStr.String()
+}
+
+// addInitializationGraphOp Add the initialization operation to the graph to
+// cover all the added variables
+func (gr *Graph) addInitializationGraphOp() {
+	inputs := make([]string, len(gr.variables))
+	i := 0
+	for input := range gr.variables {
+		inputs[i] = "^" + input + "/Assign"
+		i++
+	}
+
+	gr.def.Node = append(gr.def.Node, &pb.NodeDef{
+		Name:  "init",
+		Op:    "NoOp",
+		Input: inputs,
+	})
+}
+
+// Placeholder Adds a placeholder to the Graph, a placeholder is an
+// operation that must be fed with data on execution.
+func (gr *Graph) Placeholder(name string, dataType DataType, dims []int64, dimNames []string) (op *GraphNode) {
+	op = &GraphNode{
+		outDataTypes: map[string]DataType{
+			name: dataType,
+		},
+		def: &pb.NodeDef{
+			Name: name,
+			Op:   "Placeholder",
+			Attr: make(map[string]*pb.AttrValue),
+		},
+	}
+	op.def.Attr["dtype"] = &pb.AttrValue{
+		Value: &pb.AttrValue_Type{
+			Type: pb.DataType(dataType),
+		},
+	}
+
+	shape := &pb.TensorShapeProto{
+		Dim: make([]*pb.TensorShapeProto_Dim, len(dims)),
+	}
+
+	for i, dim := range dims {
+		shape.Dim[i] = &pb.TensorShapeProto_Dim{
+			Size: dim,
+		}
+
+		if len(dimNames) == len(dims) {
+			shape.Dim[i].Name = dimNames[i]
+		}
+	}
+
+	op.def.Attr["shape"] = &pb.AttrValue{
+		Value: &pb.AttrValue_Shape{
+			Shape: shape,
+		},
+	}
+
+	gr.def.Node = append(gr.def.Node, op.def)
+
+	return
+}
+
+// AsStr Returns the current graph serialized so it can be exported.
+func (gr *Graph) AsStr() []byte {
+	result, _ := proto.Marshal(gr.def)
+
+	return result
+}
+
+// ErrExpectedVarAsinput The specified operation is not defined.
+type ErrExpectedVarAsinput struct {
+	operation string
+	inputPos  int
+}
+
+func (e *ErrExpectedVarAsinput) Error() string {
+	return fmt.Sprintf(
+		"The expected input value at pos %d for the operation '%s' must be of type Variable",
+		e.inputPos, e.operation)
+}
+
+// ErrOperationNotFound The specified operation is not defined.
+type ErrOperationNotFound struct {
+	op string
+}
+
+func (e *ErrOperationNotFound) Error() string {
+	return fmt.Sprintf("Operation '%s' not defined", e.op)
+}
+
+// ErrInvalidAmounthOfInputs The number of inputs doesn't corresponds with the
+// expected for this operation.
+type ErrInvalidAmounthOfInputs struct {
+	operation  string
+	opInputs   int
+	specInputs int
+}
+
+func (e *ErrInvalidAmounthOfInputs) Error() string {
+	return fmt.Sprintf("Inputs required for operation '%s': %d, but %d provided",
+		e.operation, e.opInputs, e.specInputs)
+}
+
+// ErrMandatoryAttributeNotSpecified A mandatory attribute for this operation
+// was not specified.
+type ErrMandatoryAttributeNotSpecified struct {
+	operation  string
+	attribName string
+}
+
+func (e *ErrMandatoryAttributeNotSpecified) Error() string {
+	return fmt.Sprintf("The attribute '%s' is mandatory for the operation: '%s'",
```

Comment:
I'd change "the operation" -> "operation" for consistency with the other error messages.
Also, only this and the next error have a colon after operation: '%s'.  The previous ones don't.  I'd remove it and just say 'after operation '%s'.  The use throughout the rest of your error messages uses the colon to separate the error and an explanation, and that seems more consistent.